code/espurna/telnet.ino

```diff
@@ -120,13 +127,101 @@ void _telnetData(unsigned char clientId, void *data, size_t len) {
     #if TERMINAL_SUPPORT
         terminalInject(data, len);
     #endif
+}
+
+void _telnetNotifyConnected(unsigned char i) {
+
+    DEBUG_MSG_P(PSTR("[TELNET] Client #%u connected\n"), i);
+
+    // If there is no terminal support automatically dump info and crash data
+    #if TERMINAL_SUPPORT == 0
+        info();
+        wifiDebug();
+        crashDump();
+        crashClear();
+    #endif
+
+    #ifdef ESPURNA_CORE
+        _telnetClientsAuth[i] = true;
+    #else
+        _telnetClientsAuth[i] = !_telnetAuth;
+        if (_telnetAuth) {
+            if (getAdminPass().length()) {
+                _telnetWrite(i, "Password: ");
+            } else {
+                _telnetClientsAuth[i] = true;
+            }
+        }
+    #endif
+
+    _telnetFirst = true;
+    wifiReconnectCheck();
 
 }
 
-void _telnetNewClient(AsyncClient *client) {
+#if TELNET_SERVER == TELNET_SERVER_WIFISERVER
 
-    if (client->localIP() != WiFi.softAPIP()) {
+void _telnetLoop() {
+    if (_telnetServer.hasClient()) {
+        int i;
+
+        for (i = 0; i < TELNET_MAX_CLIENTS; i++) {
+            if (!_telnetClients[i] || !_telnetClients[i]->connected()) {
+
+                _telnetClients[i] = std::unique_ptr<WiFiClient>(new WiFiClient(_telnetServer.available()));
+
+                if (_telnetClients[i]->localIP() != WiFi.softAPIP()) {
+                    // Telnet is always available for the ESPurna Core image
+                    #ifdef ESPURNA_CORE
+                        bool telnetSTA = true;
+                    #else
+                        bool telnetSTA = getSetting("telnetSTA", TELNET_STA).toInt() == 1;
+                    #endif
+
+                    if (!telnetSTA) {
+                        DEBUG_MSG_P(PSTR("[TELNET] Rejecting - Only local connections\n"));
+                        _telnetDisconnect(i);
+                        return;
+                    }
+                }
+
+                _telnetNotifyConnected(i);
+
+                break;
+            }
+        }
 
+        //no free/disconnected spot so reject
+        if (i == TELNET_MAX_CLIENTS) {
+            DEBUG_MSG_P(PSTR("[TELNET] Rejecting - Too many connections\n"));
+            _telnetServer.available().stop();
+            return;
+        }
+    }
+
+    for (int i = 0; i < TELNET_MAX_CLIENTS; i++) {
+        if (_telnetClients[i]) {
+            // Handle client timeouts
+            if (!_telnetClients[i]->connected()) {
+                _telnetDisconnect(i);
+            } else {
+                // Read data from clients
+                while (_telnetClients[i] && _telnetClients[i]->available()) {
+                    size_t len = _telnetClients[i]->available();
+                    char data[len];
```

Comment:
Returning to my previous comment regarding stack, I don't think it makes sense to have this dynamic.
What I missed was that old 512 might've been too big too. _telnetData(...) will feed the data[] to the terminal buffer and it is limited by 128 bytes (TERMINAL_BUFFER_SIZE) anyways

Reply:
OK, changed now